Wallet setup redesign

dennisreimann opened the pull request:
WIP for #1703 and #1755.

As the wallet setup wizard is quite encapsulated, we're also using this as a start in the new design direction. E.g. added an icon sprite for the new icons made by @dstrukt and I also want to implement the new checkbox style (that switch thingy).

Building this in parallel to the existing derivation scheme workflow and I'm trying to reuse as much as possible from that. Not focussing too much on the design details for now, as the finetuning is still in progress and we can do that using this PR as a prototype once the basic workflow is done.

dennisreimann commented:
@dstrukt @pavlenex Some notes about edge cases that we need to incorporate: 

- Setup: The option to create a wallet might not be available. The server admin has a setting for that (Server settings: Policies). 
- Import: The "Connect hardware wallet" is only available with JavaScript enabled, because the Vault Bridge needs it.

I propose we show the options nevertheless, but instead of the link/button to proceed we display an explanation about why this isn't available.

dennisreimann commented:
The hardware wallet setup flow works now. Will do the other ones for existing wallets next, then creating a new wallet. Afterwards the tests might get fixed too. 😉

dennisreimann commented:
Alright, tests are working besides one flakyintegration test (`CanGetRateCryptoCurrenciesByDefault`) – marking this for review ready now, as it is pretty big already.

**Please do not merge it before the imminent next release**, because I see this as part 1 of what needs to be done.

Once this gets merged, there are some more things to do:

- Remove the old setup flow, with controllers and views (I've build the new one alongside it)
- Refactor the new flow to integrate some of the private controller methods from the controller that will be removed
- Refactor the current scan modal so that the functionality can be used inline in the scan page as well
- Discuss and decide on the [open points mentioned above](https://github.com/btcpayserver/btcpayserver/pull/2164#issuecomment-751997058) 
- Finalize the design according to @dstrukt's mockups

NicolasDorier commented:
Normally there is a retry button to this page, but I don't see it here.

![image](https://user-images.githubusercontent.com/3020646/103894504-60d94b00-5132-11eb-911d-e9e2f295065e.png)

NicolasDorier commented:
![image](https://user-images.githubusercontent.com/3020646/103894658-94b47080-5132-11eb-98d8-7f36697f6fb8.png)

Actually myself I don't even know about this one, so some text would be nice.

NicolasDorier commented:
![image](https://user-images.githubusercontent.com/3020646/103894810-c9282c80-5132-11eb-9452-cf9e0839b18a.png)

We should say it is the BIP39 password.

![image](https://user-images.githubusercontent.com/3020646/103894870-dc3afc80-5132-11eb-9228-bb85673facd8.png)

Can you change the text to "anyone with admin access to this server"

NicolasDorier commented:
We can't disable the payment method anymore?

NicolasDorier commented:
`Account` is an advanced setting.

![image](https://user-images.githubusercontent.com/3020646/103895368-a9453880-5133-11eb-86b9-51b973bd9ea6.png)

dennisreimann commented:
> We can't disable the payment method anymore?

@NicolasDorier Good point and it's one of the open points from above that I'd like to discuss. imho the solution like we had it before doesn't work well with this wizard style anymore. I'd rather make this option available on the update store page as you also suggested before (when we rebrushed the wallet display on that page). Having a toggle there would work better than going through thee setup wizard to turn it off/on.

NicolasDorier commented:
@dennisreimann yeah, I wanted to do that for long. To keep this PR short, you can do it only for on-chain. For lightning we can do in next release.

dennisreimann commented:
Keep it short. Yeah, that's what she said …

<img width="127" alt="commit" src="https://user-images.githubusercontent.com/886/103897046-3caa4980-50f3-11eb-9df6-2d03cd3eb45a.png">

dennisreimann commented:
@NicolasDorier Addressed your points. Regarding the enable/disable toggle I've contacted @dstrukt to figure out what a good solution might look like.

dstrukt commented:
All great points brought up, and for the most part aware of. I have a few improvements and thoughts to address them as well.

@dennisreimann and I will be meeting tomorrow to chat progress so I can reign in the remaining edge case and close the loop on the wizard UX.

Will follow up with updates.

dstrukt commented:
@dennisreimann and I just met, and decided on the last few bits of functional elements remaining to reign in this PR.

I'll be doing one final pass over the UI in Figma to get everything ready for the following PR.

Once I've got that last pass done, @pavlenex the three of us (+ potentially others) can take a look over a few specific screens to finalize the wording on each of the sections, so Dennis only has to do the UI polish pass once.

Otherwise, I think we're in a great place!

pavlenex commented:
@dennisreimann @dstrukt is there ETA for this though? We're looking to release a new version early next week, so if it takes more than that, we should probably leave it for the next one.

dennisreimann commented:
@pavlenex see [my comment above](https://github.com/btcpayserver/btcpayserver/pull/2164#issuecomment-756011458) – this won't be ready for the next release as I consider this the first of two PRs that should be done before releasing it.

pavlenex commented:
Ok, we can postpone it for the next release then. But @Kukks had the idea of not removing legacy stuff in the first release of the wallet setup. @Kukks can you elaborate on the reasoning behind this idea?

dennisreimann commented:
Maybe it's related to altcoin wallet setup? How does that work?

pavlenex commented:
From what I remember, @Kukks suggested that we should have the old way of wallet setup in there at leats for a while until we're 100% sure that the new one works well.

Kukks commented on BTCPayServer/Views/Stores/ImportWallet/Scan.cshtml:
```diff
@@ -0,0 +1,46 @@
+                document.getElementById("qr-import-form").submit();
+            }, "scanqrModal");
+
+            $("#scanqrModal").modal();
```
We should hook into the modal's close event and make a redirect to "back" when it closes.  Alternatively, we can just modify the content of the modal and remove the backdrop and X button from it so that it is just embedded into the page (kind of)

dennisreimann replied:
I plan to do the latter in an additional PR. I've chosen not to do it here because this PR is very big already and we can do this in a more distinct polishing PR after merging this one.

Kukks commented on BTCPayServer/Models/StoreViewModels/WalletSetupViewModel.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    public enum WalletSetupMethod
+    {
+        Undefined, // needs to be present and first to cover the null case
```
You could just have `WalletSetupMethod?` as a variable?

dennisreimann replied:
You mean the `public WalletSetupMethod Method { get; set; }` like it's defined below?

Kukks replied:
yeah and then you can remove the `Undefined` option

Kukks reviewed (changes requested):
Looks great so far, already a big improvement!

Some points:
* We lost the enable/disable option
* Cannot view an already set-up wallet. Perhaps it can be the confirm addresses page + the wallet xpub to not add more pages?
* altcoins need to have Vault hidden away
* Im not sure if it's still there, but altcoins still allowed the legacy ledger support

dennisreimann commented:
@Kukks 

> We lost the enable/disable option
> Cannot view an already set-up wallet.

@dstrukt is currently on it: We plan to introduce a separate page you get to for the "Modify" case. Right now you land in the setup process again, which isn't ideal – so this will be replace with a page where you have the option to 1) enable/disable, 2) change the settings (redo the setup) or 3) remove the wallet settings.

> altcoins need to have Vault hidden away

The [vault option is only present for BTC](https://github.com/btcpayserver/btcpayserver/pull/2164/files#diff-cbe1530074a0350bc6370ce010ad59cda34bbca249fc999c828eb2ceb12ced3fR20).

> Im not sure if it's still there, but altcoins still allowed the legacy ledger support

This should still be present, but I'll check it.

dennisreimann commented on BTCPayServer/Views/Stores/ImportWallet/Xpub.cshtml:
```diff
@@ -0,0 +1,81 @@
+        </thead>
+        <tbody>
+        <tr>
+            <td>P2WPKH</td>
```
I think we should. This is an advanced/expert user option and they should know what this means. Can you provide my with the updates for the table?

dennisreimann commented on BTCPayServer/Views/Stores/SetupWallet.cshtml:
```diff
@@ -0,0 +1,42 @@
+        {
+            <div class="list-group-item p-4 text-muted">
+                <h4>Create a new wallet</h4>
+                <p class="mb-0">Please note that creating a wallet is not supported by your instance.</p>
```
Done 👍

dennisreimann commented:
Added the modify variant as a mockup and work towards what @dstrukt might suggest.

The page contains the current wallet settings as well as the necessary options to 
1. enable/disable the wallet
2. change the settings by redoing the setup
3. remove the wallet entirely (instead of just disabling it)

@kukks @NicolasDorier For the removal I am unsure if what I did is enough. Can you [verify this code is correct](https://github.com/btcpayserver/btcpayserver/pull/2164/commits/64a4df4142d5c40fc70bf3fca263841a0481ceba#diff-cf9136a24f0d17d58e06a6c27f55b6454bf3492aebb03556a0c48c6b24e1032bR476)? 

![modify](https://user-images.githubusercontent.com/886/104216969-0e05d900-543b-11eb-82af-87c58a40905b.png)

dennisreimann commented:
> Im not sure if it's still there, but altcoins still allowed the legacy ledger support

@Kukks is it this legacy address type option you meant?

![legacy](https://user-images.githubusercontent.com/886/104218098-8d47dc80-543c-11eb-9857-ae2a028a56c4.png)

dstrukt commented:
"Added the modify variant as a mockup and work towards what @dstrukt might suggest."

Indeed, this is the jist of what we discussed on friday for modification of an established wallet view. Nice to see it all in one place... will clean in the Figma.

pavlenex commented:
Continued testing this, here are some notes:

1. **Wallet Modify** - It's unclear what's the distinction between `Wallet > Settings` and `stores/xyz/wallet/BTC/modify` From the text `Change your current wallet settings` and CTA 
`Change wallet-Update settings, changes take effect on setup completion`, it is extremely hard and confusing to distinguish and understand the difference between the two. Should they be the same? When I click `Change` wallet it leads me to the initial setup.
2. `Air-gapped import by scanning a QR code shown on your wallets display` - This does not seem to be correct. What this really does is imports a scanned xpub. I am thinking if we really want this displayed this prominently, since it's only supported with what BlueWallet and Cobo? Also description should be `Import by scanning a wallet a QR code of your extended public key (Supported only by Blue Wallet and XYZ) - may require a link to docs in the future.
3. In `Generate a new wallet` the CTA is `Continue` which indicates that there is one more step, I suggest we change it to `Create a wallet` though this isn't critical.

Besides this, there are some textual changes that we should improve to make things more obvious.

Really an amazing progress great work @dennisreimann and @dstrukt :rocket:

Kukks commented:
> Continued testing this, here are some notes:
> 
> 1. **Wallet Modify** - It's unclear what's the distinction between `Wallet > Settings` and `stores/xyz/wallet/BTC/modify` From the text `Change your current wallet settings` and CTA
>    `Change wallet-Update settings, changes take effect on setup completion`, it is extremely hard and confusing to distinguish and understand the difference between the two. Should they be the same? When I click `Change` wallet it leads me to the initial setup.
> 2. `Air-gapped import by scanning a QR code shown on your wallets display` - This does not seem to be correct. What this really does is imports a scanned xpub. I am thinking if we really want this displayed this prominently, since it's only supported with what BlueWallet and Cobo? Also description should be `Import by scanning a wallet a QR code of your extended public key (Supported only by Blue Wallet and XYZ) - may require a link to docs in the future.
> 3. In `Generate a new wallet` the CTA is `Continue` which indicates that there is one more step, I suggest we change it to `Create a wallet` though this isn't critical.
> 
> Besides this, there are some textual changes that we should improve to make things more obvious.
> 
> Really an amazing progress great work @dennisreimann and @dstrukt 🚀

I would leave the camera option there but we may want to change the copy

dennisreimann commented:
Yep, copy and design finetuning need to be done. I'll continue with extracting the QR scan from the modal.

rockstardev commented:
There is a small bug with navigation in wallet wizard when modifying:


https://user-images.githubusercontent.com/5191402/104853866-3b360980-58c9-11eb-9135-73679e1d51c0.mp4

britttttk commented:
When modifying an existing wallet (in this example I have: xpub from QR code > changing to wallet from seed)

issues:
1. If the user adds additional spaces in the seed string it will create a new seed for the user
2. If the user realizes this is not right and tries to navigate back from the "confirm new seed page", the wallet is already saved.


https://user-images.githubusercontent.com/39231115/104994425-a83cc280-59e1-11eb-850b-1f10b6fbaabb.mp4

Eskyee commented:
Good find.. newbies don't understand a  space can create a different seed!! And the wallet saved!! Not a good look.. if they lose access

dstrukt commented:
Ahhh agreed, nice find -- this is probably the area least refined/discussed thus far, we definitely need to think about how to potentially mitigate these basic/expected user errors.

dennisreimann commented:
The navigation bugs are fixed.

Regarding the seed bug that Britt mentioned I'll have to take a detailed look tomorrow, because it looks like something funky is going on there and a new walet is generated regardless of additional spaces or not. Will also check the additional spaces stuff once the general bug is fixed.

@dstrukt Though it doesn't seem perfect to me, I'll also tweak the modify screen in this direction you proposed on the Mattermost chat. Looks cleaner than what we have currently, but we'll have to discuss the details in the meeting on friday. Re-posting that screen here to entertain the discussion here :)

![Screen Shot 2021-01-18 at 1 45 03 AM](https://user-images.githubusercontent.com/886/105214821-cc6ae180-5b50-11eb-89c3-55dade07e87d.png)

dstrukt commented:
Don't worry about making it match exactly, that is still very much WIP design to help direct the discussion, attempting to group / minimize the UI and choices on this screen.

Sounds good, I think we should be able to iron out the remaining issues on friday.

dennisreimann commented:
The seed import is also fixed now. It wasn't about the spaces – I messed up the flow so that after the confirm address page a new wallet got generated regardless of the previuous input. Thanks for spotting that, @britttttk.

Also updated the modify page.

Now that the tests are also passing we can talk about everything in the design meeting tomorrow and get this finetuned, so it can hopefully be part of the next release :)

pavlenex commented:
[Notes from Design call #2](https://github.com/btcpayserver/organization/issues/16)

- Separate hot wallet vs watch-only wallet setup
- RPC wallet needs clarification ping Kukks to clarify and decide if we want this
- Account Settings remove when generating a new wallet
- Remove green separators
- Scan wallet QR code – add details about wallet support possibly in the scan tab as we do on the wallet file.
- Modify page when `enable` is clicked communicate it / Add new buttons 
- Enable/Disable on modify should be buttons

dennisreimann commented:
Updated the copy and instructions on the import pages. Still not perfect, but @pavlenex and I decided to focus on finishing the technical tasks first and enhance this after the merge.

![import](https://user-images.githubusercontent.com/886/106278319-d5d40800-623a-11eb-9d06-de4a7452f40f.png)

---

![file](https://user-images.githubusercontent.com/886/106278391-f13f1300-623a-11eb-9ddd-e356977e42bd.png)

---

![scan](https://user-images.githubusercontent.com/886/106278342-df5d7000-623a-11eb-9ee9-11de3f4ca3ac.png)

---

![seed](https://user-images.githubusercontent.com/886/106278351-e2f0f700-623a-11eb-9fbb-fa27ec666d9f.png)

dennisreimann commented:
As discussed in last weeks design meeting I've now split the options for generating a new wallet: We now distinguish the different use cases of a hot and watch-only wallet to also communicate the differences.

This isn't completely done yet and the tests also need fixing, but I wanted to post this to get the discussion going. So here are the screens of the flow: (adding the start screen for the sake of completeness)

![start](https://user-images.githubusercontent.com/886/106311346-48f27400-6265-11eb-9a2c-d6eb4de01fd9.png)

## Generating a new wallet: The options

![generate-options](https://user-images.githubusercontent.com/886/106311405-60c9f800-6265-11eb-9bbb-4c616ee285ca.png)

There's also one question I have: When the `CanUseHotWallet` flag isn't set – does it mean one just cannot generate a host wallet or one cannot generate a wallet at all? I didn't really grasp it from the previous flow … maybe @Kukks or @NicolasDorier can answer this one :)

However, this is what it looks like right now …

![generate-options-disableed](https://user-images.githubusercontent.com/886/106311403-60316180-6265-11eb-8229-b213ee2e5af1.png)

## Generating a new wallet: The forms

The RPC toggle isn't on by default, just wanted to show the different states …

![generate-hotwallet](https://user-images.githubusercontent.com/886/106311578-aab2de00-6265-11eb-9187-4da6e471b49f.png)

---

![generate-watchonly](https://user-images.githubusercontent.com/886/106311590-ad153800-6265-11eb-9e75-c1f69db69e1d.png)


## Import Seed

Posting this one, because it also got updated in relation to the screens above …

![import-seed](https://user-images.githubusercontent.com/886/106311697-d766f580-6265-11eb-9c6e-08fd81e358c6.png)

NicolasDorier commented:
My merge finger start itching

dennisreimann commented:
@NicolasDorier When the CanUseHotWallet flag isn't set – does it mean one just cannot generate a hot wallet or one cannot generate a wallet at all? I didn't really grasp it from the previous flow …

dstrukt commented:
I love to see this progress, will be working this weekend to get the remaining bits cleaned up so we're ready for the follow up PR.

Excellent work @dennisreimann!!

Kukks commented:
> @NicolasDorier When the CanUseHotWallet flag isn't set – does it mean one just cannot generate a hot wallet or one cannot generate a wallet at all? I didn't really grasp it from the previous flow …

Technically it is only really meant to not allow hot wallets stored on the server. BUT the current UI just hides the generate wallet entirely

dennisreimann commented:
I think this is done from a technical perspective. Everything that has been discussed is in here and the tests are also working now. So I think it can be reviewed now. 😬

bolatovumar reviewed (approved):
Tested locally on a Mac and looks really nice!

Some minor issues I noticed:

For some reason SVG icons were not loading in Opera and Firefox at first, I had to refresh the page a few times for them to show up. In Chrome they showed up right away on the first try.
![Screen Shot 2021-02-04 at 6 45 30 PM](https://user-images.githubusercontent.com/1934678/106983565-f5779e80-671a-11eb-9ff8-0bc4d3e7f213.png)

The icons and heading can overlap a little on smaller screens:
![Screen Shot 2021-02-04 at 6 48 22 PM](https://user-images.githubusercontent.com/1934678/106983672-25bf3d00-671b-11eb-9ba8-d0d30c38aee7.png)
![Screen Shot 2021-02-04 at 6 49 56 PM](https://user-images.githubusercontent.com/1934678/106983704-3374c280-671b-11eb-826e-16d388512895.png)

This table doesn't show up full-width on smaller screens but shows up full-width on larger screens:
![Screen Shot 2021-02-04 at 6 51 41 PM](https://user-images.githubusercontent.com/1934678/106983722-3bccfd80-671b-11eb-881f-bbae4715247a.png)

When using the Vault option it would be nice if it told you to check your Vault app and accept the connection. I didn't know I had to do it at first and was confused as to why it wasn't finding my Vault and was just stuck in this state. Also, if you reject the connection in Vault the page will be stuck in this state (not sure if there is a way to know if user has rejected the connection from Vault):
![Screen Shot 2021-02-04 at 6 54 36 PM](https://user-images.githubusercontent.com/1934678/106983766-50a99100-671b-11eb-82d1-e060ba9b9533.png)

dennisreimann commented:
@bolatovumar thanks a lot, this is exactly the kind of feedback I need to proceed here! 👍

dennisreimann commented:
Addressed @bolatovumar's points. As for the Vault: The message after rejecting the connections is working for me – this is what it looks like:

![vault-reject](https://user-images.githubusercontent.com/886/107049622-9c674380-67ca-11eb-9d29-9d7b3b33e420.png)

bolatovumar commented:
> Addressed @bolatovumar's points. As for the Vault: The message after rejecting the connections is working for me – this is what it looks like:
> 
> ![vault-reject](https://user-images.githubusercontent.com/886/107049622-9c674380-67ca-11eb-9d29-9d7b3b33e420.png)

I will take a look again, I probably did something wrong.

bolatovumar reviewed (approved):
Tried it again and when I reject the Vault connection I get the appropriate message in the UI now. Not sure why I wasn't getting it previously.

![Screen Shot 2021-02-07 at 2 19 07 PM](https://user-images.githubusercontent.com/1934678/107161320-e1b57d80-6950-11eb-9ea8-dd042d57cf4f.png)

I did notice though that if I reject it and then try again a few times it can get out of sync and I stop getting the message in the UI.

Kukks commented on BTCPayServer/Views/Stores/ImportWallet/File.cshtml:
```diff
@@ -0,0 +1,54 @@
+}
+
+@section Navbar {
+    <a asp-controller="Stores" asp-action="ImportWallet" asp-route-storeId="@Model.StoreId" asp-route-cryptoCode="@Model.CryptoCode" asp-route-method="">
```
why is there an empty route method attr?

dennisreimann replied:
Without it, the framework seems to use the current route value, so that navigating back isn't possible here … we need to explicitely unset it.

Kukks commented on BTCPayServer/Views/Stores/ImportWallet/File.cshtml:
```diff
@@ -0,0 +1,54 @@
+    </tr>
+    <tr>
+        <td>Electrum</td>
+        <td>File ❯ Save backup</td>
```
Maybe we should mention that it must NOT be encrypted with a password

Kukks commented on BTCPayServer/Views/Stores/ImportWallet/Xpub.cshtml:
```diff
@@ -0,0 +1,74 @@
+    </tr>
+    </thead>
+    <tbody>
+    <tr>
```
We should mention that 
* `ypub...` (aka P2SH-P2WPKH) and 
* `zpub...` (P2WPKH) 
* output descriptors:
  * `pkh([d34db33f/44'/0'/0']xpub.../0/*)` (P2PKH and with the account key settings configured) 
  * `pkh(xpub.../0/*)` (P2PKH)
  * `sh(wpkh([d34db33f/49'/0'/0']xpub.../0/*)` (P2SH-P2WPKH and with the account key settings configured) 
  *  `sh(wpkh(xpub../0/*)` (P2SH-P2WPKH)
  *  `wpkh([d34db33f/84'/0'/0']xpub.../0/*)` (P2WPKH and with the account key settings configured) 
  *  `wpkh(xpub/0/*)` (P2WPKH) 
  * `sh(multi(1,[d34db33f/45'/0]xpub.../0/*,[d34db33f/45'/0]xpub.../0/*))` (Multi-sig P2SH)
  * `wsh(multi(1,[d34db33f/48'/0'/0'/2']xpub../0/*,[d34db33f/48'/0'/0'/2']xpub.../0/*))`
 are also supported (Multi-sig P2WSH)
  * `sh(wsh(multi(1,[d34db33f/48'/0'/0'/1']xpub../0/*,[d34db33f/48'/0'/0'/1']xpub.../0/*)))` (Multi-sig P2SH-P2WSH)
  * sortmulti variant for multisig 

are also supported

dennisreimann replied:
Added, thanks a lot!

![addrtypes](https://user-images.githubusercontent.com/886/107357357-3e8c7180-6ad2-11eb-9b9c-573205a64cd9.png)

Kukks commented on BTCPayServer/Views/Stores/ImportWalletOptions.cshtml:
```diff
@@ -0,0 +1,123 @@
+                            <span class="mr-2">Connect hardware wallet</span>
+                            <span class="badge bg-primary">Recommended</span>
+                        </h4>
+                        <p class="mb-0">Please enable JavaScript for this option to be available</p>
```
We can hide this automatically using `hide-when-js` when JS is already enabled.

dennisreimann replied:
It's inside a `noscript` block, so it won't be shown when JS is enabled. It's kind of an `else` for the linked block above it.

Kukks commented on BTCPayServer/Views/Stores/ImportWalletOptions.cshtml:
```diff
@@ -0,0 +1,123 @@
+        <div class="content">
+            <h4>
+                <span class="mr-2">Enter wallet seed</span>
+                <span class="badge bg-danger">Not recommended</span>
```
Maybe a tooltip explaining why?

dstrukt replied:
Noted, will bring up in tomorrow's discussion.

dennisreimann replied:
Done in 044629861006c1080d001aa2c7222a9f03d86227.

![tooltip](https://user-images.githubusercontent.com/886/107383294-2841de00-6af1-11eb-9dd9-abf08a539c18.png)

Kukks commented on BTCPayServer/Views/Stores/_GenerateWalletForm.cshtml:
```diff
@@ -0,0 +1,123 @@
+        </button>
+        <div id="advanced-settings" class="collapse @(string.IsNullOrEmpty(Model.Passphrase) && !Model.ImportKeysToRPC ? "" : "show")">
+            <div class="pt-3 pb-1">
+                @if (isImport) // hide account option when creating a wallet
```
Why hide the account option?

pavlenex replied:
Because it's not useful when we create a new wallet, it just adds another confusion step?

dennisreimann replied:
Yep, that's the reasoning: When creating a new wallet we just go with the first account …

Kukks replied:
It is in the advanced section though? It is almost equivalent to not allowing a user to specify their passphrase (as a completely different wallet is generated)

dennisreimann replied:
I'd vote for keeping it simple as I guess 99% will most likely go with the first account anyways. Nevertheless, let's discuss this in #2277 as I might be overlooking something here.

Kukks reviewed (commented):
Went through code but will ned to run it and check it out too

dstrukt commented:
Beautiful feedback @bolatovumar & @Kukks - thank you!

Will address some of the visual feedback in tomorrow's call.

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+        [HttpPost]
+        [Route("{storeId}/wallet/{cryptoCode}/modify")]
+        [Route("{storeId}/wallet/{cryptoCode}/import/{method}")]
+        [ApiExplorerSettings(IgnoreApi = true)]
```
It's a leftover from the BTCLike partial controller I used as a basis … will remove it, thanks for spotting!

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+                                : new BitcoinExtPubKey(vm.AccountKey, network.NBitcoinNetwork);
+                            if (accountKey != null)
+                            {
+                                var accountSettings =
```
It's the [same as before](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Controllers/StoresController.BTCLike.cs#L143), because I used the BTCLike Controller as a basis.

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+            {
+                try
+                {
+                    if (!string.IsNullOrEmpty(vm.DerivationScheme))
```
It's the [same as before](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Controllers/StoresController.BTCLike.cs#L135). I tried to refrain from changing logic as much as possible – let's do those kinds of refactorings once this is merged, ok?

dennisreimann replied:
Done in https://github.com/btcpayserver/btcpayserver/pull/2296/commits/240ab37a67dcbb1158795be30e12fc9ab52bf611

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+            if (!ModelState.IsValid || !(result is RedirectToActionResult))
+                return result;
+
+            TempData.Clear();
```
Good spot!

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+            }
+
+            var wallet = _WalletProvider.GetWallet(network);
+            if (wallet == null)
```
Done in 5f74334bbd3a7dbd8fc2256938b9731455972153.

dennisreimann commented on BTCPayServer/Controllers/StoresController.BTCWallet.cs:
```diff
@@ -0,0 +1,557 @@
+                return NotFound();
+            }
+
+            return View("Confirm", new ConfirmModel
```
Done in df7b2d88bef990989c6ca5852379d965525ae563.

![warning](https://user-images.githubusercontent.com/886/107383228-13fde100-6af1-11eb-8d7a-f60427bf7add.png)

pavlenex commented:
It's up for the final review, @NicolasDorier this PR is in your hands now :rocket:

dennisreimann commented:
Rebased … `CanCreateApiKeys` fails unrelatedly. 

@kukks Are you [ok with this](https://github.com/btcpayserver/btcpayserver/pull/2164#discussion_r572718221)? If so, I think @NicolasDorier can merge this tomorrow.

NicolasDorier commented:
testing

NicolasDorier commented:
For later PRs:

![image](https://user-images.githubusercontent.com/3020646/107626954-e310de00-6ca1-11eb-9a43-30f2388a1f63.png)

We should mention fully noded or put a link to documentation explaining what you can do when addresses are imported in the bitcoin node.

Damn the interface is slick

pavlenex commented:
Woohooo :rocket: :rocket: :rocket:

dennisreimann commented:
Awesome, thanks to everyone who helped with this – @dstrukt and @pavlenex for concept and design, @britttttk  and @bolatovumar for testing, @kukks and @NicolasDorier for review and suggestions! This is a great team effort and we can be proud of the result 🚀

MaxHillebrand commented:
Major work - congrats & thanks all!
